### PR TITLE
fix: schema error when parsing order-by expressions

### LINKED ISSUE #10013: HAVING doesn't work with ORDER BY
### Describe the bug

I found a case where a query fails when there is both a HAVING clause and an ORDER BY clause. The query will work if it has either the HAVING clause or the ORDER BY clause .. but not both.

Here is a failing query:
```sql
with revs
as (
    SELECT
        customer,
        cast(round(CAST(substring(revenue,2) AS float)) as INT) AS visit_revenue
    FROM visits
)
SELECT
        customer,
        SUM(visit_revenue) as total_revenue
FROM revs
group by customer
having SUM(visit_revenue) > 999
order by SUM(visit_revenue)
```
The error I get is: 
```shell
Error: SchemaError(FieldNotFound { field: Column { relation: None, name: "visit_revenue" }, valid_fields: [Column { relation: Some(Bare { table: "revs" }), name: "customer" }, Column { relation: None, name: "total_revenue" }, Column { relation: Some(Bare { table: "revs" }), name: "customer" }, Column { relation: None, name: "SUM(revs.visit_revenue)" }] }, Some(""))
```

If I comment out either the line with `having SUM(visit_revenue) > 999` OR `order by SUM(visit_revenue)` the the query will succeed.

### To Reproduce

To reproduce, run this program with the attached CSV file as data.

```rust
use datafusion::arrow::datatypes::{DataType, Field, Schema};
use datafusion::prelude::{CsvReadOptions, SessionContext};

#[tokio::main]
async fn main() -> datafusion::error::Result<()> {
    let file_name = "data/test.csv";

    let schema = Schema::new(vec![
        Field::new("date", DataType::Utf8, false),
        Field::new("weekday", DataType::Utf8, false),
        Field::new("time", DataType::Utf8, false),
        Field::new("customer", DataType::Utf8, false),
        Field::new("revenue",  DataType::Utf8, false),
    ]);

    let options = CsvReadOptions::new()
        .schema(&schema)
        .delimiter(b'\t')
        .has_header(true)
        ;

    let ctx = SessionContext::new();
    ctx.register_csv("visits", file_name, options).await?;

    let df = ctx.sql(r#"
                with revs
                as (
                    SELECT
                        customer,
                        cast(round(CAST(substring(revenue,2) AS float)) as INT) AS visit_revenue
                    FROM visits
                )
                SELECT
                        customer,
                        SUM(visit_revenue) as total_revenue
                FROM revs
                group by customer
                having SUM(visit_revenue) > 999
                order by SUM(visit_revenue)
                "#).await?;

    df.show().await?;
    Ok(())
}
```
The attached file contains the data:
[test.csv](https://github.com/apache/arrow-datafusion/files/14918379/test.csv)

### Expected behavior

The expectation is that the result set is both filtered for records where the sum of `visit_revenue` > 999 AND ordered by the same without crashing. 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Closes #10013.

## Rationale for this change
In the following query syntax, `SELECT select_list FROM table_expression ORDER BY expressions`, the order-by expressions can not only reference the column names in the select list, but also reference the column names in the `FROM` clause. 

Currently, when [building](https://github.com/apache/datafusion/blob/f8c623fe045d70a87eac8dc8620b74ff73be56d5/datafusion/sql/src/expr/order_by.rs#L33) order-by expressions, only the input plan's schema (derived from the select list) is used.
For the following query, this will cause the column reference `a` in `ORDER BY SUM(a)` to fail to [normalize](https://github.com/apache/datafusion/blob/f8c623fe045d70a87eac8dc8620b74ff73be56d5/datafusion/expr/src/expr_rewriter/order_by.rs#L103).
```sh
DataFusion CLI v37.0.0
> create table t(a int);
0 row(s) fetched.
Elapsed 0.005 seconds.

> SELECT
   SUM(a)
FROM t
having SUM(a) > 0
order by SUM(a);
Schema error: No field named a. Valid fields are "SUM(t.a)", "SUM(t.a)".
```

The solution is, when constructing order-by expressions, we can use both the schema of the select list and the schema of the `FROM` clause. 

To achieve this, we need to handle the `ORDER BY` clause within the planning of select, that is, `select_to_plan`.
This approach may also have other benefits:
- Refine the current [add_missing_columns](https://github.com/apache/datafusion/blob/f8c623fe045d70a87eac8dc8620b74ff73be56d5/datafusion/expr/src/logical_plan/builder.rs#L437) logic by directly adding the missing columns to the select list, which should be more efficient and less susceptible to errors.
- Easier to coordinate `DISTINCT` and `ORDER BY`, because distinct is defined within the select list.
- Easier to expand the functionality of `ORDER BY`, such as supporting order by unprojected aggregate expressions and [unprojected window functions](https://github.com/apache/datafusion/issues/608). The term 'unprojected' means that they do not appear in the select list.

These can be implemented in subsequent PRs.

## What changes are included in this PR?
Use both the schema of the select list and the FROM clause to construct order-by expressions.

## Are these changes tested?
Yes.
By existing tests and new tests.

## Are there any user-facing changes?
No

### CODE REVIEW

**Comment by alamb** [comment]
Thank you for this PR @jonahgao -- it is on my review list for tomorrow

**Review by comphead — COMMENTED** [review]
thanks @jonahgao for addressing this problem. Im thinking if we can avoid of duplicating schema, this is usually super confusing. Perhaps we can construct or amend the schema before calling the order by?

**Comment by jonahgao** [comment]
> thanks @jonahgao for addressing this problem. Im thinking if we can avoid the additional schemas in the signature, this is usually super confusing, we recently removed the similar from windows. Perhaps we can construct or amend the schema before calling the order by?

Here we need to distinguish these two schemas. When the order by expression is a literal, such as `ORDER BY 1`, it can only refer to the input schema, that is, the literal is the index of the select list.

**Review by alamb — COMMENTED** [review]
TLDR thank you @jonahgao  -- while I also have some concerns about this PR as described below, given it fixes a bug I think we could merge it as is. However, I think it might be good to have a broader discussion about this

> Currently, when [building](https://github.com/apache/datafusion/blob/f8c623fe045d70a87eac8dc8620b74ff73be56d5/datafusion/sql/src/expr/order_by.rs#L33) order-by expressions, only the input plan's schema (derived from the select list) is used.
For the following query, this will cause the column reference a in ORDER BY SUM(a) to fail to [normalize](https://github.com/apache/datafusion/blob/f8c623fe045d70a87eac8dc8620b74ff73be56d5/datafusion/expr/src/expr_rewriter/order_by.rs#L103).

I feel like I am missing something in this explanation. The following query works without this PR (and shows that ORDER BY can reference columns from the FROM clause, not just what is in the SELECT list)

```sql
> create table foo(x int, y int);
0 row(s) fetched.
Elapsed 0.002 seconds.

> select x from foo order by y;
+---+
| x |
+---+
+---+
0 row(s) fetched.
Elapsed 0.019 seconds.
```

The only type of query that this PR seems to solve involves the `HAVING` clause -- maybe the issue is that the schema used to to resolve the HAVING clause needs to be treated like the ORDER BY clause?

File: datafusion/sqllogictest/test_files/order.slt [hunk 1]
```diff
@@ -888,6 +888,30 @@ select column2, column1 from foo ORDER BY column2 desc, column1 desc;
 [1] 4
 [0] 2
 
+# Test issue: https://github.com/apache/datafusion/issues/10013
+# There is both a HAVING clause and an ORDER BY clause in the query.
+query I
+SELECT
+   SUM(column1)
+FROM foo
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/order.slt:896]: Could we also add an example that doesn't have a HAVING as well as one with GROUP BY? something like this perhaps

```sql
SELECT SUM(column1) FROM foo ORDER BY SUM(column1)
```

```sql
SELECT column2 FROM foo ORDER BY SUM(column1)
```

```sql
SELECT SUM(column1) FROM foo ORDER BY SUM(column1) GROUP BY column2
```

**Review by alamb — APPROVED** [review]

File: datafusion/sqllogictest/test_files/order.slt [hunk 2]
```diff
@@ -888,6 +888,30 @@ select column2, column1 from foo ORDER BY column2 desc, column1 desc;
...
+HAVING SUM(column1) > 0
+ORDER BY SUM(column1)
+----
+16
+
+statement ok
+create table ambiguity_test(a int, b int) as values (1,2), (2,1);
```
[the remark this answers is not in the document]
Comment (jonahgao) [reply datafusion/sqllogictest/test_files/order.slt]: Fixed.

File: datafusion/sqllogictest/test_files/order.slt (hunk 1 above)
Comment (jonahgao) [reply datafusion/sqllogictest/test_files/order.slt:896]: Added. Thank you @alamb !

**Comment by jonahgao** [comment]
> The following query works without this PR (and shows that ORDER BY can reference columns from the FROM clause, not just what is in the SELECT list)

@alamb  `select x from foo order by y` can be covered by [add_missing_columns](https://github.com/apache/datafusion/blob/f8c623fe045d70a87eac8dc8620b74ff73be56d5/datafusion/expr/src/logical_plan/builder.rs#L437),  by blindly adding columns into the descendant projection node. Another issue is that  we should not run `add_missing_columns`  for other `SetExpr`s except SELECT.
```sh
DataFusion CLI v37.1.0
> create table t(a int, b int);

> select a from t union select 1 order by b;
Error during planning: For SELECT DISTINCT, ORDER BY expressions b must appear in select list
> select a from t union all select 1 order by b;
Schema error: No field named t.a. Valid fields are a, t.b.
```
Doing it for UNION makes the error messages hard to understand.
[PostgreSQL](https://www.postgresql.org/docs/current/queries-order.html) refuses to do this, order by can only references output columns. 
> ORDER BY can be applied to the result of a UNION, INTERSECT, or EXCEPT combination, but in this case it is only permitted to sort by output column names or numbers.

**Comment by jonahgao** [comment]
> The only type of query that this PR seems to solve involves the HAVING clause -- maybe the issue is that the schema used to to resolve the HAVING clause needs to be treated like the ORDER BY clause?

@alamb  I think we should make the column qualified, not the other way around. The problem of the `SUM()` example is that it can't resolve the qualifier of `a` in `ORDER BY SUM(a)`, to match the select list item which has been normalized into `SUM(t.a)`. 

I think that we should handle `ORDER BY` similarly to `HAVING`, use the merged schema, add the missing columns directly in the select list, instead of traversing the plan looking for projection node. Their processing logic may be reusable.

**Comment by alamb** [comment]
> I think that we should handle `ORDER BY` similarly to `HAVING`, use the merged schema, add the missing columns directly in the select list, instead of traversing the plan looking for projection node. Their processing logic may be reusable. I agree it might be good to have a broader discussion about this.

Yes, this makes sense -- so in this case perhaps it would mean removing [add_missing_columns](https://github.com/apache/datafusion/blob/f8c623fe045d70a87eac8dc8620b74ff73be56d5/datafusion/expr/src/logical_plan/builder.rs#L437) and unifying with this code path.

I'll file a ticket to discuss

**Review by alamb — APPROVED** [review]
Thanks again @jonahgao  and @comphead  - I think this PR is an improvement in that it fixes a bug, even though there is likely more room for code improvement (tracked in https://github.com/apache/datafusion/issues/10326)

**Comment by jonahgao** [comment]
Thanks for the review @comphead @alamb ❤️
